### PR TITLE
Add tooltips to difficulty icons with extra details

### PR DESCRIPTION
**Pull request by GSculerlor**
Supersedes #4794. Now the design matched with web version 💃

Current result:
Sample
Another sample

- [x] Depends on #5744.

### CODE REVIEW

**Comment by peppy** [comment]
I think I would want it to display everywhere, including direct.

**Comment by GSculerlor** [comment]
I mean something like in here. I'm pretty sure tooltip is redundant here since we already have view for difficulty version and star rating and it's block the view
example

**Comment by peppy** [comment]
I see. You could make a local nested class there which overrides the tooltip to null, to forcefully hide it.

**Comment by GSculerlor** [comment]
Just check the `TooltipContainer` on framework, it say that if the `IHasCustomTooltip`'s `GetCustomTooltip()` is null, it'll use the default tooltip. So if I override to make tooltip null, it just make the tooltip back to default one

**Comment by peppy** [comment]
I think you can also provide `string.Empty` via `IHasTooltip` to get around that?

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs [hunk 1]
```diff
@@ -23,14 +23,14 @@ public class DifficultyIcon : DifficultyColouredContainer, IHasCustomTooltip
     {
         private readonly RulesetInfo ruleset;
 
-        public DifficultyIcon(BeatmapInfo beatmap, RulesetInfo ruleset = null)
+        public DifficultyIcon(BeatmapInfo beatmap, RulesetInfo ruleset = null, Boolean shouldShowTooltip = false)
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/DifficultyIcon.cs]: wouldn't it be best to make this true by default, since that's the most common use case?

**Comment by peppy** [comment]
image

This looks wrong.

Also the background colour doesn't match normal tooltips. Please update the OsuTooltip to use the style you've implemented (if the one you implemented is deemed more correct / in line with web).

**Comment by peppy** [comment]
I'd go as far as making a base class for `DifficultyTooltip` and `OsuTooltip` to share (basically sharing the background, and transition logic, including the instant-move portion of `OsuTooltip`).

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs [hunk 2]
```diff
@@ -1,36 +1,127 @@
...
+                        RelativeSizeAxes = Axes.Both
+                    },
+                    new FillFlowContainer
+                    {
+                        AutoSizeAxes = Axes.Both,
+                        Direction = FillDirection.Vertical,
+                        Padding = new MarginPadding(10),
+                        Children = new Drawable[]
+                        {
+                            difficultyName = new OsuSpriteText
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/DifficultyIcon.cs]: What happens if the difficulty name is very long?

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs [hunk 3]
```diff
@@ -1,36 +1,127 @@
...
+
+            [BackgroundDependencyLoader]
+            private void load(OsuColour colours)
+            {
+                background.Colour = colours.GreyCarmineDark;
+            }
+
+            public void Refresh()
+            {
+                var info = TooltipText.Split('$');
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/DifficultyIcon.cs]: This is really awkward, but I can see why it was done this way. `TooltipContainer` is a bit awkward.

**Comment by peppy** [comment]
This also needs tests. Including edge cases which would test long difficulty names and also test every colour combination to catch things like I found above.

**Comment by GSculerlor** [comment]
Yeah, already created test locally to check it and found that awkward string split can be wrong when it found diff with dollar sign like https://osu.ppy.sh/beatmapsets/101766#osu/272026

Will continue this probably by tomorrow

**Comment by peppy** [comment]
You'll also find that very long titles will likely make hugely horizontal tooltips, that can potentially exit the screen.

**Comment by GSculerlor** [comment]
> You'll also find that very long titles will likely make hugely horizontal tooltips, that can potentially exit the screen.

Yes can confirm it
Example

Also about the background since the web is dark (should be align with mine), I need to change the default one to apply same colour?

**Comment by iiSaLMaN** [comment]
> > You'll also find that very long titles will likely make hugely horizontal tooltips, that can potentially exit the screen.
> 
> Yes can confirm it
> Example
> 
> Also about the background since the web is dark (should be align with mine), I need to change the default one to apply same colour?

A simple fix for that is limiting the text with `...`

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs (hunk 3 above)
Comment (GSculerlor) [reply osu.Game/Beatmaps/Drawables/DifficultyIcon.cs]: For now will change the separator with another symbol. Should I need framework side change for this?

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs (hunk 1 above)
Comment (iiSaLMaN) [reply osu.Game/Beatmaps/Drawables/DifficultyIcon.cs]: `Boolean` can be simplified into `bool`

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs [hunk 4]
```diff
@@ -1,36 +1,127 @@
...
+                                Anchor = Anchor.Centre,
+                                Origin = Anchor.Centre,
+                                Font = OsuFont.GetFont(size: 16, weight: FontWeight.Bold),
+                            },
+                            new FillFlowContainer
+                            {
+                                AutoSizeAxes = Axes.Both,
+                                Anchor = Anchor.Centre,
+                                Origin = Anchor.Centre,
+                                Direction = FillDirection.Horizontal,
```
Comment (iiSaLMaN) [inline osu.Game/Beatmaps/Drawables/DifficultyIcon.cs]:
```suggestion
                                Direction = FillDirection.Horizontal,
                                Colour = accentColour,
```
You can just set the color of this container to the accent color instead of manually doing that for each child

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs [hunk 5]
```diff
@@ -1,36 +1,127 @@
...
+            TooltipText = shouldShowTooltip ? $"{beatmap.Version}${beatmap.StarDifficulty:0.##}" : String.Empty;
 
             Size = new Vector2(20);
         }
 
+        public string TooltipText { get; set; }
+
+        public ITooltip GetCustomTooltip() => new DifficultyIconTooltip(AccentColour);
+
+        public class DifficultyIconTooltip : VisibilityContainer, ITooltip
```
Comment (iiSaLMaN) [inline osu.Game/Beatmaps/Drawables/DifficultyIcon.cs]:
```suggestion
        private class DifficultyIconTooltip : VisibilityContainer, ITooltip
```
Should be private since the only usage for it is inside DifficultyIcon

**Review by frenzibyte — CHANGES_REQUESTED** [review]
Small changes

**Comment by peppy** [comment]
This still needs a bit more thinking as to how to make custom tooltips behave more like the default one. I'll look at making framework changes to support this as necessary.

**Review by smoogipoo — DISMISSED** [review]
LGTM for now :)

**Review by smoogipoo — APPROVED** [review]

**Comment by peppy** [comment]
I guess we can fix the colour clash for expert in a follow-up PR.
